### PR TITLE
feature: non-trivial instance construction

### LINKED ISSUE #28: Non-trivial construction
Right now, the materializer is akin to:

``` csharp
TheType obj = new();
while (...)
{
    switch (...)
    {
        case 4223423:
            obj.Foo = /* read the value */
            break;
    }
}
return obj;
```

This only works for trivial construction; it does not support more complex scenarios:

1. custom constructors
2. init-only properties
3. read-only fields (deferred for now, as this also requires materializer colocation)

---

A custom constructor (at most one per type) is identified by:

- ignore parameterless constructors and constructors that are marked `[DapperAot(false)]`
- if only a single constructor remains, it is selected whether or not it is marked `[DapperAot]`/`[DapperAot(true)]`
- if multiple constructors remain, and multiple are marked `[DapperAot]`/`[DapperAot(true)]`, a generator error is emitted and no constructor is selected
- if multiple constructors remain, and exactly one is marked `[DapperAot]`/`[DapperAot(true)]`, it is selected
- in all other cases, no constructor is selected

If a custom constructor is selected, the parameters are checked against the member list using normalized / insensitive rules; if any paired members do not match field type (question: nullability?), a generator error is emitted and the constructor is deselected

---

Init-only properties are any non-static properties that have the `init` modifier, for example `public string Name { get; init; }`

---

In any of these scenarios, we must collect the values into locals, and defer construction; I would propose simply `value0`, `value1`, etc where the index is the position of the discovered member; each should be assigned `default`, awaiting values from the database; we then collect fields into these variables instead of the object:

``` csharp
int value0 = default;
string? value1 = default;
DateTime value2 = default;

while (...)
{
    switch (...)
    {
        case 4223423:
            value1 = /* read the value */
            break;
    }
}
// construction not shown yet
```

---

The construction step depends on whether a custom constructor is selected, and whether the parameters for such custom constructor  covers all members; there are 3 possible outcomes:

1. custom constructor covers all members

``` csharp
return new TheType(value0, value2, value1);
```

(noting that the parameter order does not necessarily match our declaration order, so it is not necessarily strict order)

2. custom constructor covers some but not all members (which may or may not include `init`-only properties)

``` csharp
return new TheType(value0, value2)
{
    Foo = value1,
};
```

3. no custom constructor (and which by elimination: must include at least one `init`-only property)

``` csharp
return new TheType
{
    Foo = value1,
};
```

---

Implementation notes:

DRY:

- ideally the 3 non-trivial construction techniques (with/without constructor, with/without additional members) should not use 3 separate paths (it is fine to keep the original simple construction `TheType obj = new()` line separate, for simplicity; see †)
- ideally the field read loop should not be duplicated between simple/custom construction; it should just change the target of the read statement between `obj.TheMember =` vs `value42 =`

†: the 4th quadrant in that with/without matrix is: simple construction, so: already handled - i.e. without constructor, everything is additional members, none of which are `init`-only - the difference being that in that simple scenario, we're not using the stack to gather the values - we're pushing them directly onto the object

### PR DESCRIPTION
**Pull request by DeagleGross**
Closes #28

### CODE REVIEW

File: src/Dapper.AOT.Analyzers/CodeAnalysis/DapperInterceptorGenerator.cs [hunk 1]
```diff
@@ -1167,24 +1169,59 @@ private static void WriteCommandFactory(string baseFactory, CodeWriter sb, IType
...
+            .Where(member => CodeWriter.IsSettableInstanceMember(member.Member, out _))
+            .ToImmutableArray();
+        var membersCount = members.Length;
+        if (membersCount == 0)
         {
-            if (CodeWriter.IsSettableInstanceMember(member.Member, out _))
-            {
-                memberCount++;
-            }
+            // report diagnostic? Or do we want to allow returning empty instance like `return new UserType();` ?
```
Comment (DeagleGross) [inline src/Dapper.AOT.Analyzers/CodeAnalysis/DapperInterceptorGenerator.cs:1180]: @mgravell, please, review this diagnostic report. I dont think it makes sense to generate "empty" `RowFactory` for the user type, which does not have any settable members (which means that only empty instance can be returned from Dapper).

Let me know if that makes sense, or I need to rollback this change

File: src/Dapper.AOT.Analyzers/CodeAnalysis/DapperInterceptorGenerator.cs [hunk 2]
```diff
@@ -13,6 +13,8 @@
 using System.Collections.Immutable;
 using System.Data;
 using System.Linq;
+using System.Reflection;
+using System.Runtime.InteropServices.ComTypes;
```
Comment (mgravell) [inline src/Dapper.AOT.Analyzers/CodeAnalysis/DapperInterceptorGenerator.cs:17]: double-check this maybe? is this a "devenv being devenv" thing?

File: src/Dapper.AOT.Analyzers/CodeAnalysis/DapperInterceptorGenerator.cs [hunk 3]
```diff
@@ -1167,24 +1169,58 @@ private static void WriteCommandFactory(string baseFactory, CodeWriter sb, IType
...
+            .Where(member => CodeWriter.IsSettableInstanceMember(member.Member, out _))
+            .ToImmutableArray();
+        var membersCount = members.Length;
+        if (membersCount == 0)
         {
-            if (CodeWriter.IsSettableInstanceMember(member.Member, out _))
-            {
-                memberCount++;
-            }
+            context.ReportDiagnostic(Diagnostic.Create(Diagnostics.UserTypeNoSettableMembersFound, type.Locations.First()));
```
Comment (mgravell) [inline src/Dapper.AOT.Analyzers/CodeAnalysis/DapperInterceptorGenerator.cs:1180]: should this logic only happen after resolving a constructor? if there's a constructor, the members needn't be settable...

File: src/Dapper.AOT.Analyzers/Internal/Inspection.cs [hunk 4]
```diff
@@ -1,11 +1,15 @@
-﻿using Dapper.Internal.Roslyn;
+﻿using Dapper.CodeAnalysis;
+using Dapper.Internal.Roslyn;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.SqlServer.TransactSql.ScriptDom;
```
Comment (mgravell) [inline src/Dapper.AOT.Analyzers/Internal/Inspection.cs:5]: double-check whether this is needed

File: test/Dapper.AOT.Test/Interceptors/QueryCustomConstruction.input.cs [hunk 5]
```diff
@@ -0,0 +1,127 @@
+﻿using Dapper;
+using System.Data.Common;
+
+[module: DapperAot]
+
+public static class Foo
```
Comment (mgravell) [inline test/Dapper.AOT.Test/Interceptors/QueryCustomConstruction.input.cs:6]: propose scenario:

- get-only properties via contructor
- record class
- record struct

File: src/Dapper.AOT.Analyzers/CodeAnalysis/DapperInterceptorGenerator.cs (hunk 2 above)
Comment (DeagleGross) [reply src/Dapper.AOT.Analyzers/CodeAnalysis/DapperInterceptorGenerator.cs]: that is me being not careful, sorry! fixed

File: src/Dapper.AOT.Analyzers/Internal/Inspection.cs (hunk 4 above)
Comment (DeagleGross) [reply src/Dapper.AOT.Analyzers/Internal/Inspection.cs]: removed, thanks for noticing!

File: src/Dapper.AOT.Analyzers/CodeAnalysis/DapperInterceptorGenerator.cs (hunk 3 above)
Comment (DeagleGross) [reply src/Dapper.AOT.Analyzers/CodeAnalysis/DapperInterceptorGenerator.cs]: Yep, did not think about this scenario. Reporting diagnostic, if members **and** constructor is not found.
`membersCount == 0 && !hasDapperAotCompatibleConstructor`

File: test/Dapper.AOT.Test/Interceptors/QueryCustomConstruction.input.cs (hunk 5 above)
Comment (DeagleGross) [reply test/Dapper.AOT.Test/Interceptors/QueryCustomConstruction.input.cs:6]: supported get-only properties via ctor

File: src/Dapper.AOT.Analyzers/CodeAnalysis/DapperInterceptorGenerator.cs (hunk 1 above)
Comment (DeagleGross) [reply src/Dapper.AOT.Analyzers/CodeAnalysis/DapperInterceptorGenerator.cs]: resolving - seems like it's OK

File: test/Dapper.AOT.Test/Interceptors/QueryCustomConstruction.input.cs (hunk 5 above)
Comment (DeagleGross) [reply test/Dapper.AOT.Test/Interceptors/QueryCustomConstruction.input.cs:6]: and supported short syntax of ctors for `record`s and `record struct`s:
```
public record RecordClassSimpleCtor(int X, string Y, double? Z);
public record struct RecordStructSimpleCtor(int X, string Y, double? Z);
```

File: test/Dapper.AOT.Test/Interceptors/GlobalFetchSize.output.txt [hunk 6]
```diff
@@ -1,4 +1,7 @@
-Generator produced 1 diagnostics:
+Generator produced 2 diagnostics:
 
 Hidden DAP000  L1 C1
 Dapper.AOT handled 1 of 1 enabled call-sites using 1 interceptors, 1 commands and 1 readers
+
+Error DAP037 Interceptors/GlobalFetchSize.input.cs L18 C18
+Type has none of settable members (fields or properties)
```
[the remark this answers is not in the document]
Comment (DeagleGross) [reply test/Dapper.AOT.Test/Interceptors/GlobalFetchSize.output.txt:7]: Error tells you that there is no point in doing a `Query<Entity>`, when `Entity` does not have members (properties\fields). Take a look at this example:
```
public class MyEntity
{
   // nothing here
}

public async Task RunQuery()
{
   using var connection = new SqlConnection();
   var result = await connection.Query<MyEntity>("my sql");
   // result will always be just an empty object: no data inside
}
```

We can either remove such diagnostic, or try to fix the formulation of error. What would you like?

[the remark this answers is not in the document]
Comment (mgravell) [reply test/Dapper.AOT.Test/Interceptors/GlobalFetchSize.output.txt:7]: Just a wording thing, then? maybe "Type has no settable members (fields or properties)"

[the remark this answers is not in the document]
Comment (DeagleGross) [reply test/Dapper.AOT.Test/Interceptors/GlobalFetchSize.output.txt]: Changed. Now the reported error is `Type 'SomeApp.SomeQueryType' has no settable members (fields or properties)`

File: test/Dapper.AOT.Test/Interceptors/QueryCustomConstruction.output.cs [hunk 7]
```diff
@@ -0,0 +1,956 @@
...
+                        break;
+                    case 5:
+                        value2 = reader.IsDBNull(columnOffset) ? (double?)null : GetValue<double>(reader, columnOffset);
+                        break;
+
+                }
+                columnOffset++;
+
+            }
+            return new global::Foo.GetOnlyPropertiesViaConstructor(value0,value1,value2);
```
Comment (mgravell) [inline test/Dapper.AOT.Test/Interceptors/QueryCustomConstruction.output.cs:319]: nit: spacing - `(value0, value1, value2)`

**Review by mgravell — COMMENTED** [review]
looking good; tiny nit on the spacing - petty but worth fixing; I suggest we merge after that, but a possible follow-up / extension (separate PR) would be to maybe take the same idea but generalized to include factory methods, i.e. `[DapperAot] public static Foo Create(int id, DateTime when, string whatever)` - very similar rules, but different nuances - for example, we can't support init-only members after a factory, but we *can* still support settable props, i.e.

``` c#
var obj = Foo.Create(value0, value2, value1);
obj.Blah = value4;
```

thoughts?

File: test/Dapper.AOT.Test/Interceptors/QueryCustomConstruction.output.cs (hunk 7 above)
Comment (DeagleGross) [reply test/Dapper.AOT.Test/Interceptors/QueryCustomConstruction.output.cs]: fixed

**Comment by DeagleGross** [comment]
> looking good; tiny nit on the spacing - petty but worth fixing; I suggest we merge after that, but a possible follow-up / extension (separate PR) would be to maybe take the same idea but generalized to include factory methods, i.e. `[DapperAot] public static Foo Create(int id, DateTime when, string whatever)` - very similar rules, but different nuances - for example, we can't support init-only members after a factory, but we _can_ still support settable props, i.e.
> 
> ```cs
> var obj = Foo.Create(value0, value2, value1);
> obj.Blah = value4;
> ```
> 
> thoughts?

sure, created an issue for follow-up: #34

**Pull request merged**